### PR TITLE
add dismount volume during UnmountDevice for windows

### PR DESCRIPTION
**Pull request by jingxu97**
The main purpose of this PR is to support NTFS on Linux. It requires following changes

- dismount volume for Windows during volume tear down phase. 
The reason for dismount volume on Windows is because the file system in an “unclean” state without a propose dismount. After volume is detached from Windows node, it will cause issues when mounting it on to a Linux node next.

- With dismount added, drive letter cannot be used. 
Before, durning format and mount, drive letter is automatically assigned. But after dismount, when disk is attached again, it will fail to mount because drive letter is unassigned. 
https://github.com/kubernetes/utils/pull/179 is merged.
Need to update k/k vendor for utils.

### CODE REVIEW

**Comment by jingxu97** [comment]
pull-kubernetes-e2e-kind

**Comment by jingxu97** [comment]
pull-kubernetes-e2e-kind

**Comment by jingxu97** [comment]
pull-kubernetes-e2e-kind

**Comment by jingxu97** [comment]
pull-kubernetes-e2e-kind

**Comment by jingxu97** [comment]
cc @msau

**Comment by jingxu97** [comment]
@verult @msau42 PTAL

**Comment by jingxu97** [comment]
@verult @msau42 PTAL

File: pkg/volume/gcepd/attacher.go [hunk 1]
```diff
@@ -393,10 +397,42 @@ func (detacher *gcePersistentDiskDetacher) UnmountDevice(deviceMountPath string)
...
+	cmd := fmt.Sprintf("(Get-Item %s).Target", path)
+	output, err := exec.Command("powershell", "/c", cmd).CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("get path %s target failed: %v, output: %q", path, err, string(output))
+	}
+	target := string(output)
+	if len(target) < 1 {
+		return "", fmt.Errorf("get device path for %s failed, output is empty", target)
+	}
+	target = strings.Replace(target, "\r\n", " ", -1)
```
Comment (verult) [inline pkg/volume/gcepd/attacher.go:432]: is the space in `" "` intentional?

**Review by verult — COMMENTED** [review]
LGTM from my perspective.

Is there / will there be a test covering this?

**Comment by msau42** [comment]
Can you describe the issue this is fixing?

File: pkg/volume/gcepd/attacher.go (hunk 1 above)
Comment (jingxu97) [reply pkg/volume/gcepd/attacher.go]: removed it.

**Comment by jingxu97** [comment]
> Can you describe the issue this is fixing?

Added comments about this PR. PTAL

File: pkg/volume/gcepd/attacher.go [hunk 2]
```diff
@@ -393,10 +402,42 @@ func (detacher *gcePersistentDiskDetacher) UnmountDevice(deviceMountPath string)
...
+	cmd := fmt.Sprintf("(Get-Item %s).Target", path)
+	output, err := exec.Command("powershell", "/c", cmd).CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("get path %s target failed: %v, output: %q", path, err, string(output))
+	}
+	target := string(output)
+	if len(target) < 1 {
+		return "", fmt.Errorf("get device path for %s failed, output is empty", target)
+	}
+	target = strings.Replace(target, "\r\n", "", -1)
```
Comment (msau42) [inline pkg/volume/gcepd/attacher.go:437]: Would `strings.TrimSpace` or any of those variants work better?

File: pkg/volume/gcepd/attacher.go [hunk 3]
```diff
@@ -393,10 +402,42 @@ func (detacher *gcePersistentDiskDetacher) UnmountDevice(deviceMountPath string)
 		if err != nil {
 			return err
 		}
+		target, err := getPathTarget(deviceMountPath)
+		if err != nil {
+			return err
+		}
+		if err = dismount(target); err != nil {
```
Comment (msau42) [inline pkg/volume/gcepd/attacher.go:409]: is there a reason why we only do this for gce pd plugin and not do it as a normal part of Unmount?

**Review by msau42 — COMMENTED** [review]
> Need to update k/k vendor for utils.

Does this mean we should not merge this until k/k is updated with the drive letter fix?

File: pkg/volume/gcepd/attacher.go (hunk 3 above)
Comment (jingxu97) [reply pkg/volume/gcepd/attacher.go:409]: dismount here should be only performed as part of unmountDevice before detaching. If multiple pods mount the same disk on the same node, dismount will break other pods accessing the disk.

**Comment by jingxu97** [comment]
@msau42 PTAL

File: pkg/volume/gcepd/attacher.go (hunk 3 above)
Comment (msau42) [reply pkg/volume/gcepd/attacher.go:409]: Is this specific to pd, or can other block storage drivers use it like ebs, azure block?

File: pkg/volume/gcepd/attacher.go (hunk 2 above)
Comment (jingxu97) [reply pkg/volume/gcepd/attacher.go]: replaced.

**Comment by jingxu97** [comment]
> > Need to update k/k vendor for utils.
> 
> Does this mean we should not merge this until k/k is updated with the drive letter fix?

yes, i think so

File: pkg/volume/gcepd/attacher.go (hunk 3 above)
Comment (jingxu97) [reply pkg/volume/gcepd/attacher.go:409]: I think it is not specific to pd.

**Comment by jingxu97** [comment]
@msau42 I think all comments were addressed. PTAL.

File: pkg/volume/gcepd/attacher.go (hunk 3 above)
Comment (msau42) [reply pkg/volume/gcepd/attacher.go:409]: I think we should follow up and try to get this logic into the mount library because it's not specific to pd.

**Review by msau42 — COMMENTED** [review]
/lgtm
/approve
/hold
to wait for the driver letter fix

**Review by jingxu97 — COMMENTED** [review]
> /lgtm
> /approve
> /hold
> to wait for the driver letter fix

I checked the master, the driver letter fix is already in.

File: pkg/volume/gcepd/attacher.go (hunk 3 above)
Comment (jingxu97) [reply pkg/volume/gcepd/attacher.go:409]: Create bug to track it https://github.com/kubernetes/kubernetes/issues/93978

**Comment by msau42** [comment]
But there's a problem with the drive letter fix with subpath?

**Comment by jingxu97** [comment]
> But there's a problem with the drive letter fix with subpath?

right, then we can wait for that fix is merged so that we can test this one then.

**Pull request closed without merging**